Readme.md
# idb-factory

> Better window.indexedDB.

[![](https://saucelabs.com/browser-matrix/idb-factory.svg)](https://saucelabs.com/u/idb-factory)

[![](https://img.shields.io/npm/v/idb-factory.svg)](https://npmjs.org/package/idb-factory)
[![](https://img.shields.io/travis/treojs/idb-factory.svg)](https://travis-ci.org/treojs/idb-factory)
[![](http://img.shields.io/npm/dm/idb-factory.svg)](https://npmjs.org/package/idb-factory)

This module provides consistent, modern API to `window.indexedDB`.
It's especially useful for test environment, when you need to open/delete database multiple times.
For implementation details see the [150 lines of the source](./src/index.js).

## Example

```js
import { open, del } from 'idb-factory'

// open database with version 1 and create stores
open('mydb', 1, upgradeCallback).then((db) => {
  expect(db.version).equal(2)
  // use db ...
  // delete database
  return del(db)
})

function upgradeCallback(e) {
  e.target.result.createObjectStore('books', { keyPath: 'id' })  
  e.target.result.createObjectStore('magazines')  
}
```

## API

`open` and `del` each return a `Promise`. See the section on blocking
events for more details.

### open(dbName, [version], [upgradeCallback])

```js
import { open } from 'idb-factory'

(async () => {

// open "mydb1" v1, and create store and index
const db1 = await open('mydb1', 1, (e) => {
  if (e.oldVersion < 1) {
    const store = e.target.result.createObjectStore('books', { keyPath: 'isbn' })
    store.createIndex('by_title', 'title', { unique: true })
  }  
})

// version and upgradeCallback are optional.
const db2 = await open('mydb2')

})();
```

### del(db)

```js
import { open, del } from 'idb-factory'

(async () => {
const db = await open('mydb')

// do something with db
// ...

// delete existing IDBDatabase instance
await del(db)

})();
```

### del(dbName)

```js
import { del as deleteDatabase } from 'idb-factory'

(async () => {
await deleteDatabase('mydb') // delete database by name

})();
```

Note that due to a [browser bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1220279)
the `oldVersion` property of `del` error events does not work properly in
Firefox, but we have at least polyfilled its problem with `newVersion`
(which should be `null`).

We've similarly polyfilled its `success` event (for the sake of PhantomJS).

### cmp(val1, val2)

```js
import { cmp } from 'idb-factory'

// Compare 2 values, using IndexedDB's comparison algorithm
console.assert(cmp('z', 'a') === 1)
console.assert(cmp([1], [1]) === 0)
```

### global.forceIndexedDB

It is a special global variable, which can be defined to prior `global.indexedDB`.

```js
function idb() {
  return global.forceIndexedDB
      || global.indexedDB
      || global.webkitIndexedDB
      || global.mozIndexedDB
      || global.msIndexedDB
      || global.shimIndexedDB
}
```

For example, due to [WebKit bug](https://bugs.webkit.org/show_bug.cgi?id=137034) you can't rewrite
`window.indexedDB`, but Safari 8 implementation is [really buggy](https://gist.github.com/nolanlawson/08eb857c6b17a30c1b26),
so you'd like to use [shim and fallback to WebSQL](https://github.com/axemclion/IndexedDBShim).

```js
import 'indexeddbshim'
import { open } from 'idb-factory'

if (isSafari8) {
  global.forceIndexedDB = global.shimIndexedDB
}

(async () => {

// use WebSQL implementation, only in Safari 8,
// and use IndexedDB in remaining browsers.
const db = await open('mydb')

})();
```

### Blocking events

Genuine `blocked` events (as with errors) can be caught, and `blocked` event
objects will also be assigned a custom `resume` property promise which can be
used similarly to the original `open` (or `del`) call (upon successful closing
of all other open connections).

```js
import { del as deleteDatabase } from 'idb-factory'
try {
  async deleteDatabase('mydb') // delete database by name
} catch (err) {
  if (err.type !== 'blocked') {
    // Handle other `err` errors here
    return
  }
  closeOpenConnections()
  try {
    async err.resume
  } catch (err) {
    // Handle errors upon resumption
  }
}
```

### Blocking events

Genuine `blocked` events (as with errors) can be caught, and `blocked` event
objects will also be assigned a custom `resume` property promise which can be
used similarly to the original `open` (or `del`) call (upon successful closing
of all other open connections).

```js
import { del as deleteDatabase } from 'idb-factory'
try {
  async deleteDatabase('mydb') // delete database by name
} catch (err) {
  if (err.type !== 'blocked') {
    // Handle other `err` errors here
    return
  }
  closeOpenConnections()
  try {
    async err.resume
  } catch (err) {
    // Handle errors upon resumption
  }
}
```

## LICENSE

[MIT](./LICENSE)


History.md
##

* Breaking change (minor): `catch` to be passed error event, not error object
* Fix: Remove repeated attempt after block event as not working in browser testing
* Fix: Polyfill delete's `newVersion` for Firefox
* Fix: Polyfill PhantomJS' `del` success event, resumed and otherwise
* Fix: `preventDefault` within error handlers to avoid bubbling of errors beyond `catch`
* Feature: Support resumption of blocked events via `resume` property
* Feature: Pass on event object to `del` `onsuccess`

## 1.0.0 / 2015-11-28

* initial release :sparkles:


test/index.js
import 'indexeddbshim'
import ES6Promise from 'es6-promise'
import { expect } from 'chai'
import { open, del, cmp } from '../src'
import * as idbFactory from '../src'

describe('idb-factory', function idbFactoryTests() {
  ES6Promise.polyfill()
  this.timeout(5000)
  const dbName = 'idb-factory'

  before(() => del(dbName))

  it('exposes IDBFactory like API', () => {
    expect(idbFactory.open).a('function')
    expect(idbFactory.del).a('function')
    expect(idbFactory.cmp).a('function')
  })

  it('opens new db', () => {
    return open(dbName, 4, upgradeCallback).then((db) => {
      expect(db.version).equal(4)
      expect([].slice.call(db.objectStoreNames)).eql(['books'])
      return del(db)
    })
  })

  it('opens existing db', () => {
    return open(dbName, 3, upgradeCallback).then((db1) => {
      return open(dbName).then((db2) => {
        expect(db1.version).equal(db2.version)
        db1.close()
        db2.close()
        return del(dbName)
      })
    })
  })

  it('deletes db', () => {
    return open(dbName, 3, upgradeCallback).then((db1) => {
      db1.close()
      return del(dbName).then((e) => {
        // expect(e.oldVersion).equal(3) // Won't work in PhantomJS: https://github.com/ariya/phantomjs/issues/14141
        expect(e.newVersion).to.be.a('null')
      })
    })
  })

  it('compares 2 values', () => {
    expect(cmp(1, 5)).equal(-1)
    expect(cmp('z', 'a')).equal(1)
  })

  it('resumes from a genuine blocked event via resume property on open', () => {
    let caught = false
    return open(dbName, 3, upgradeCallback).then((db1) => {
      return open(dbName, 4).catch(function errorCatcher(err) {
        if (err.type === 'blocked') {
          // Handle other `err` errors here
          db1.close()
          caught = true
          return err.resume
        }
      }).then(function completedOpen(db2) {
        db2.close()
        expect(caught).equal(true)
        return del(dbName)
      })
    })
  })

  it('resumes from a genuine blocked event via resume property on del (and error properties are in order)', () => {
    let caught = false
    return open(dbName, 3, upgradeCallback).then((db) => {
      return del(dbName).catch(function errorCatcher(err) {
        if (err.type === 'blocked') {
          // Handle other `err` errors here
          // expect(err.oldVersion).to.equal(3) // https://bugzilla.mozilla.org/show_bug.cgi?id=1220279
          expect(err.newVersion).to.be.a('null')
          if (db) db.close()
          caught = true
          return err.resume
        }
      }).then(function completedDelete(e) {
        expect(e.oldVersion).equal(3)
        expect(e.newVersion).to.be.a('null')
        expect(caught).equal(true)
      })
    })
  })

  function upgradeCallback(e) {
    const books = e.target.result.createObjectStore('books', { keyPath: 'id' })
    books.createIndex('byTitle', 'title', { unique: true })
    books.createIndex('byAuthor', 'author')
  }
})


src/index.js

/**
 * Open IndexedDB database with `name`.
 * Retry logic allows to avoid issues in tests env,
 * when db with the same name delete/open repeatedly and can be blocked.
 *
 * @param {String} dbName
 * @param {Number} [version]
 * @param {Function} [upgradeCallback]
 * @return {Promise}
 */

export function open(dbName, version, upgradeCallback) {
  return new Promise((resolve, reject) => {
    if (typeof version === 'function') {
      upgradeCallback = version
      version = undefined
    }
    // don't call open with 2 arguments, when version is not set
    const req = version ? idb().open(dbName, version) : idb().open(dbName)
    req.onblocked = (e) => {
      const resume = new Promise((res, rej) => {
        // We overwrite handlers rather than make a new
        //   open() since the original request is still
        //   open and its onsuccess will still fire if
        //   the user unblocks by closing the blocking
        //   connection
        req.onsuccess = (ev) => res(ev.target.result)
        req.onerror = (ev) => {
          ev.preventDefault()
          rej(ev)
        }
      })
      e.resume = resume
      reject(e)
    }
    if (typeof upgradeCallback === 'function') {
      req.onupgradeneeded = (e) => {
        try {
          upgradeCallback(e)
        } catch (err) {
          // We allow the callback to throw its own error
          e.target.result.close()
          reject(err)
        }
      }
    }
    req.onerror = (e) => {
      // Prevent default for `BadVersion` and `AbortError` errors, etc.
      // These are not necessarily reported in console in Chrome but present; see
      //  https://bugzilla.mozilla.org/show_bug.cgi?id=872873
      //  http://stackoverflow.com/questions/36225779/aborterror-within-indexeddb-upgradeneeded-event/36266502
      e.preventDefault()
      reject(e)
    }
    req.onsuccess = (e) => {
      resolve(e.target.result)
    }
  })
}

/**
 * Delete `db` properly:
 * - close it and wait 100ms to disk flush (Safari, older Chrome, Firefox)
 * - if database is locked, due to inconsistent execution of `versionchange`,
 *   try again in 100ms
 *
 * @param {IDBDatabase|String} db
 * @return {Promise}
 */

export function del(db) {
  const dbName = typeof db !== 'string' ? db.name : db

  return new Promise((resolve, reject) => {
    const delDb = () => {
      const req = idb().deleteDatabase(dbName)
      req.onblocked = (e) => {
        // The following addresses part of https://bugzilla.mozilla.org/show_bug.cgi?id=1220279
        e = e.newVersion === null || typeof Proxy === 'undefined' ? e : new Proxy(e, { get: (target, name) => {
          return name === 'newVersion' ? null : target[name]
        } })
        const resume = new Promise((res, rej) => {
          // We overwrite handlers rather than make a new
          //   delete() since the original request is still
          //   open and its onsuccess will still fire if
          //   the user unblocks by closing the blocking
          //   connection
          req.onsuccess = (ev) => {
            // The following are needed currently by PhantomJS: https://github.com/ariya/phantomjs/issues/14141
            if (!('newVersion' in ev)) {
              ev.newVersion = e.newVersion
            }

            if (!('oldVersion' in ev)) {
              ev.oldVersion = e.oldVersion
            }

            res(ev)
          }
          req.onerror = (ev) => {
            ev.preventDefault()
            rej(ev)
          }
        })
        e.resume = resume
        reject(e)
      }
      req.onerror = (e) => {
        e.preventDefault()
        reject(e)
      }
      req.onsuccess = (e) => {
        // The following is needed currently by PhantomJS (though we cannot polyfill `oldVersion`): https://github.com/ariya/phantomjs/issues/14141
        if (!('newVersion' in e)) {
          e.newVersion = null
        }

        resolve(e)
      }
    }

    if (typeof db !== 'string') {
      db.close()
      setTimeout(delDb, 100)
    } else {
      delDb()
    }
  })
}

/**
 * Compare `first` and `second`.
 * Added for consistency with official API.
 *
 * @param {Any} first
 * @param {Any} second
 * @return {Number} -1|0|1
 */

export function cmp(first, second) {
  return idb().cmp(first, second)
}

/**
 * Get globally available IDBFactory instance.
 * - it uses `global`, so it can work in any env.
 * - it tries to use `global.forceIndexedDB` first,
 *   so you can rewrite `global.indexedDB` with polyfill
 *   https://bugs.webkit.org/show_bug.cgi?id=137034
 * - it fallbacks to all possibly available implementations
 *   https://github.com/axemclion/IndexedDBShim#ios
 * - function allows to have dynamic link,
 *   which can be changed after module's initial exectution
 *
 * @return {IDBFactory}
 */

function idb() {
  return global.forceIndexedDB
      || global.indexedDB
      || global.webkitIndexedDB
      || global.mozIndexedDB
      || global.msIndexedDB
      || global.shimIndexedDB
}
